### ISSUE TITLE
Problem in training with  'channels_first' mode

### ISSUE THREAD

**Issue by biswa-iitg** [qubvel/segmentation_models#184]
Is UNET model training using ResNet backbone is possible with 'channels_first' (3,H,W) mode? I am trying and the model is not learning. The same system trains well with 'channels_last' and input shape (H,W,3).

**Comment by qubvel** [comment]
Please provide a code example

**Comment by biswa-iitg** [comment]
import os
import sys
import time
import json
import glob
import pickle
import argparse
from datetime import datetime

import warnings
warnings.filterwarnings('ignore')
os.environ["TF_CPP_MIN_LOG_LEVEL"]="2"

import numpy as np
import pandas as pd

from tqdm import tqdm
from termcolor import colored, cprint

import keras
from keras.callbacks import ModelCheckpoint, TensorBoard
from image.image import ImageDataGenerator
import matplotlib.pyplot as plt
SM_FRAMEWORK=keras
import segmentation_models as sm

from segmentation_models.segmentation_models import Unet
from segmentation_models.segmentation_models import get_preprocessing
from segmentation_models.segmentation_models.losses import bce_jaccard_loss

from keras.metrics import binary_accuracy

def ensure_dir(path):
    if not os.path.exists(path):
        os.makedirs(path)

# Runtime data augmentation
def get_train_test_augmented(X_data, Y_data, batch_size, seed=10):    
    X_train = X_data
    Y_train = Y_data

    # Image data generator distortion options
    data_gen_args = dict(rotation_range=45.,
                         width_shift_range=0.1,
                         height_shift_range=0.03,
                         shear_range=10,
                         zoom_range=0.2,
                         #brightness_range=[0.95, 1.1],
                         #horizontal_flip=True,
                         #vertical_flip=True,
                         data_format='channels_first',
                         fill_mode='constant',
                         cval=0)
                         
    # Train data, provide the same seed and keyword arguments to the fit and flow methods
    X_datagen = ImageDataGenerator(**data_gen_args)
    Y_datagen = ImageDataGenerator(**data_gen_args)
    X_datagen.fit(X_train, augment=True, seed=seed)
    Y_datagen.fit(Y_train, augment=True, seed=seed)
    X_train_augmented = X_datagen.flow(X_train, batch_size=batch_size, shuffle=True, seed=seed)
    Y_train_augmented = Y_datagen.flow(Y_train, batch_size=batch_size, shuffle=True, seed=seed)
    
       
    # combine generators into one which yields image and masks
    train_generator = zip(X_train_augmented, Y_train_augmented)
    
    return train_generator

# ------------------------- Main() --------------------------------------
def main(config):
    keras.backend.set_image_data_format('channels_first')
    sm.set_framework('keras')
    
    ensure_dir(config['save_dir'])

    BACKBONE = config['architecture'].get('BACKBONE')
    preprocess_input = get_preprocessing(BACKBONE)

    # --- load data ---
    train_dict = pickle.load(open(config['datasets']['training'], "rb"))
    val_dict   = pickle.load(open(config['datasets']['validation'], "rb"))

    x_train = preprocess_input(train_dict['frames'])
    x_val   = preprocess_input(val_dict['frames'])
    
    # -- define a model ---
    model = Unet(
        BACKBONE, 
        classes          = config['architecture']['num_classes'], 
        activation       = config['architecture']['last_activation'], 
        encoder_weights  = 'imagenet',
        decoder_block_type='transpose',
        input_shape=(3, 256, 256))
        #decoder_use_batchnorm = False)
    
    model.compile('Adam', loss=bce_jaccard_loss, metrics=['accuracy'])
#    model.compile('Adam', loss=binary_crossentropy, metrics=['accuracy'])
    
    # -- callbacks ---
    time_stamp = datetime.now().strftime('%b%d-%H-%M-%S')
    filepath   = os.path.join(config['save_dir'], "{}-{}-best.hdf5".format(config['name'], time_stamp))
    checkpoint = ModelCheckpoint(filepath, monitor='val_loss', verbose=1, save_best_only=True, mode='min')
#    checkpoint = ModelCheckpoint(filepath, monitor='val_loss', verbose=1, save_best_only=False, mode='auto', period=2)
    
    ldir       = os.path.join(config['save_dir'], "{}-log-".format(config['name']) + time_stamp)
    tboard     = TensorBoard(
        log_dir        = ldir, 
        histogram_freq = 0,  
        write_graph    = False, 
        write_images   = True
    )
    
    callbacks_list = [ checkpoint, tboard ]

    
    train_generator = get_train_test_augmented(X_data=x_train, Y_data=train_dict['targets'], batch_size=config['train'].get('bs', 2), seed=10)

    M = model.fit_generator(
        #aug.flow(x_train, train_dict['targets'], batch_size=config['train'].get('bs', 2)),
        train_generator,
        epochs          = config['train'].get('epochs', 1),        
        validation_data = (x_val, val_dict['targets']),
        steps_per_epoch = 80,
        callbacks       = callbacks_list,
        verbose         = config['train']['verbosity']
    )

    # -- Loss Curve --

    fig = plt.figure(frameon=False, figsize=(15, 7), dpi=200)

    plt.subplot(1,2,1)
    plt.plot(M.history['acc'])
    plt.plot(M.history['val_acc'])
    plt.title('IoU Score Curve')
    plt.ylabel('accuracy')
    plt.xlabel('epoch')
    plt.legend(['Train_accuracy', 'Val_accuracy'], loc='upper left')



    plt.subplot(1,2,2)
    plt.plot(M.history['loss'])
    plt.plot(M.history['val_loss'])
    plt.title('Loss Curve')
    plt.ylabel('loss')
    plt.xlabel('epoch')
    plt.legend(['train_loss', 'val_loss'], loc='upper right')

    #--- fig format --
    fig.tight_layout()

    #--- save figure --

    fig.savefig(ldir + '/' + 'IoU_loss_plot.jpg')
    plt.close()
# ------------------------- __main__-------------------------------------
if __name__ == '__main__':
    print(colored('\n----- Segmentation Training Routines -----\n', 'blue', attrs=['bold']))
    
    parser = argparse.ArgumentParser(description='training script')
    parser.add_argument('-c', '--config', default=None, type=str,
                           help='json config file path (default: None)')

    args = parser.parse_args()

    # --- validate arguments  --- #
    if args.config:
        # load config file
        config = json.load(open(args.config))
    else:
        print(colored('Configuration file needs to be specified: --config', 'red'))
        sys.exit(1)    

    main(config)    

    print(colored('\n--------------- Done ----------------', 'blue', attrs=['bold']))

# ----------------------  Done -----------------------------------------

**Comment by biswa-iitg** [comment]
@qubvel Please find the code
[train_aug_py.txt](https://github.com/qubvel/segmentation_models/files/3528229/train_aug_py.txt)

**Comment by qubvel** [comment]
I have never try to train network with `channels_first`
You can do next:
1) train network with `channels_last` and save weights
2) create same network, but  with `channels_first` and load weights  

If an output is the same from both networks, the problem might be in loss function
